File: week6_dynamic_programming2/1_maximum_amount_of_gold/knapsack.cpp
#include <iostream>
#include <vector>

using std::vector;

int optimal_weight(int capacity, const vector<int> &data) {
    size_t col = capacity + 1;
    size_t row = data.size() + 1;

    std::vector<std::vector<int>> matrix(row, std::vector<int>(col));

    for (int r = 1; r < row; r++) {
        for (int c = 1; c < col; c++) {
            matrix[r][c] = matrix[r-1][c];
            if (data[r-1] <= c) {
                int val = ( matrix[r-1][c - data[r-1]] ) + data[r-1];
                if (matrix[r][c] < val) {
                    matrix[r][c] = val;
                }
            }
        }
    }

    //    for(auto& row: matrix) {for(auto& val: row) std::cout << val << ' ';std::cout << '\n';}

    return matrix[row-1][col-1];
}

int main() {
    int n;
    int weight;
    std::cin >> weight >> n;

    vector<int> data(n);
    for (int i = 0; i < n; i++) {
        std::cin >> data[i];
    }

    int optimal_value = optimal_weight(weight, data);
    std::cout << optimal_value << std::endl;
}

File: week6_dynamic_programming2/1_maximum_amount_of_gold/knapsack.py
def optimal_weight(capacity, data):
    col = capacity + 1  # weights
    row = len(data) + 1  # values

    matrix = [[0 for n in range(0, col)] for i in range(0, row)]
    # for i in range(1, row):
    #     matrix.append([0] * col)

    for r in range(1, row):
        for c in range(1, col):
            matrix[r][c] = matrix[r - 1][c]
            if data[r - 1] <= c:
                val = (matrix[r - 1][c - data[r - 1]]) + data[r - 1]
                if matrix[r][c] < val:
                    matrix[r][c] = val

    return matrix[row - 1][col - 1]


if __name__ == '__main__':
    print("main:")
    arr = [1, 4, 8]
    W = 10
    print(optimal_weight(W, arr))
